### PR TITLE
SNOW-1055755: binary releases workflow

### PR DESCRIPTION
**Pull request by sfc-gh-mraba**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on MacOS.
   * [x] I've confirmed that my changes are working by executing CLI's commands manually on Windows.
   * [x] I've confirmed that my changes are up-to-date with the target branch.
   * [x] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
#SNOW-1055755

I've added packaging scripts at `scripts/packaging` and environment hatch env `packaging`
This changes contain scripts and resources required to build binary version of snowflake-cli using pyinstaller.
This scripts are invoked by jenkins pipelines defined in this [PR](https://github.com/snowflakedb/jenkins_utils/pull/13578)

I added new field to telemetry for distinguishing execution of binaries from PyPi

### CODE REVIEW

File: scripts/packaging/macos/Resources/snowflake-cli_get_started.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    </li>
+    <li>Execute the following command to connect to Snowflake:<br/>
+        <span class="monospace">snowsql</span><br/>
+        or double click the <span class="monospace">SnowSQL</span> application icon in the <span class="monospace">Applications</span> directory.
+    </li>
+</ol>
+<br/>
+<p>See the Snowflake documentation <a href="https://docs.snowflake.net/manuals/user-guide/snowsql.html">https://docs.snowflake.net/manuals/user-guide/snowsql.html</a> for more information.</p>
+</body>
+</html>
```
Comment (sfc-gh-turbaszek) [inline scripts/packaging/macos/Resources/snowflake-cli_get_started.html:33]: This page links to snowsql, is this intentional?

File: scripts/packaging/macos/Resources/snowflake-cli_welcome.html [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    font-family: Sans-Serif;
+}
+.monospace {
+    font-family:monospace;
+    font-weight: bold;
+}
+</style>
+<body>
+<p>
+SnowSQL(<span class="monospace">snowsql</span>) is the next generation Snowflake CLI.
```
Comment (sfc-gh-turbaszek) [inline scripts/packaging/macos/Resources/snowflake-cli_welcome.html:13]: This is the other way around right?

File: scripts/packaging/macos/Resources/snowflake-cli_welcome.html [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+</p>
+<p>
+The installer copies the <span class="monospace">SnowSQL</span> application bundle to the
```
Comment (sfc-gh-turbaszek) [inline scripts/packaging/macos/Resources/snowflake-cli_welcome.html:16]: snowsql?

File: scripts/packaging/win/snowflake_cli.wxs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    </Feature>
+    <Icon Id="icon.ico" SourceFile="scripts\packaging\win\snowflake_msi.ico" />
+    <Property Id="ARPPRODUCTICON" Value="icon.ico" />
+    <Property Id="ARPHELPLINK"
+      Value="https://docs.snowflake.com/en/developer-guide/snowflake-cli-v2/index" />
+    <Property Id="ARPNOMODIFY" Value="yes" Secure="yes" />
+    <Property Id="ARPURLINFOABOUT">https://snowflake.com/</Property>
+    <Property Id="WIXUI_INSTALLDIR" Value="TESTFILEPRODUCTDIR" />
+    <UIRef Id="WixUI_InstallDir2" />
+    <Property Id="ApplicationFolderName" Value="Snowflake SQL" />
```
Comment (sfc-gh-turbaszek) [inline scripts/packaging/win/snowflake_cli.wxs:47]: Snowflake SQL?

File: scripts/packaging/win/wxs_builder.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+root = wxs.getroot()
+snow_files = root.find(".//Component[@Id='snow.exe']/..", namespaces=ns)
+if snow_files is None:
+    raise ValueError("Component not found in the template")
+
+
+lib_files = list(LIBS.glob("**/*"))
+
+for lib_path in LIBS.glob("**/*"):
+    if lib_path.is_file():
```
Comment (sfc-gh-turbaszek) [inline scripts/packaging/win/wxs_builder.py:31]:
```
if not lib_path.is_file():
   continue
```
WDYT? Less indentation

File: scripts/packaging/build_darwin_package.sh [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+    <key>CFBundleName</key>
+    <string>SnowflakeCLI</string>
+    <key>CFBundleDisplayName</key>
+    <string>SnowflakeCLI</string>
+    <key>CFBundleIdentifier</key>
+    <string>net.snowflake.snowflake-cli</string>
```
Comment (sfc-gh-astus) [inline scripts/packaging/build_darwin_package.sh:41]: `net`?

File: scripts/packaging/macos/Resources/snowflake-cli_get_started.html [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        <span class="monospace">username = &lt;user_name&gt;</span><br/>
+        <span class="monospace">password = &lt;password&gt;</span>
+    </li>
+    <li>Execute the following command to connect to Snowflake:<br/>
+        <span class="monospace">snowsql</span><br/>
+        or double click the <span class="monospace">SnowSQL</span> application icon in the <span class="monospace">Applications</span> directory.
+    </li>
+</ol>
+<br/>
+<p>See the Snowflake documentation <a href="https://docs.snowflake.net/manuals/user-guide/snowsql.html">https://docs.snowflake.net/manuals/user-guide/snowsql.html</a> for more information.</p>
```
Comment (sfc-gh-astus) [inline scripts/packaging/macos/Resources/snowflake-cli_get_started.html:31]: Link to SnowSQL docs

File: scripts/packaging/win/wxs_builder.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+wxs = ElementTree.parse(WXS_TEMPLATE_FILE)
+root = wxs.getroot()
+snow_files = root.find(".//Component[@Id='snow.exe']/..", namespaces=ns)
+if snow_files is None:
+    raise ValueError("Component not found in the template")
+
+
+lib_files = list(LIBS.glob("**/*"))
+
+for lib_path in LIBS.glob("**/*"):
```
[the remark this answers is not in the document]
Comment (sfc-gh-astus) [reply scripts/packaging/win/wxs_builder.py:30]: `lib_files` is unused.

File: scripts/packaging/macos/Resources/snowflake-cli_get_started.html [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    <li>Add your connection information to the <span class="monospace">~/.snowsql/config</span> file:<br/>
+        <span class="monospace">accountname = &lt;account_name&gt;</span><br/>
+        <span class="monospace">username = &lt;user_name&gt;</span><br/>
+        <span class="monospace">password = &lt;password&gt;</span>
+    </li>
+    <li>Execute the following command to connect to Snowflake:<br/>
+        <span class="monospace">snowsql</span><br/>
+        or double click the <span class="monospace">SnowSQL</span> application icon in the <span class="monospace">Applications</span> directory.
+    </li>
+</ol>
```
Comment (sfc-gh-astus) [inline scripts/packaging/macos/Resources/snowflake-cli_get_started.html:29]: SnowSQL few times

File: scripts/packaging/macos/Resources/snowflake-cli_get_started.html (hunk 7 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/macos/Resources/snowflake-cli_get_started.html]: Updated content.

File: scripts/packaging/macos/Resources/snowflake-cli_get_started.html (hunk 1 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/macos/Resources/snowflake-cli_get_started.html:28]: This url redirects to docs.snowflake.com
Replaced with snowflake cli docs

File: scripts/packaging/macos/Resources/snowflake-cli_welcome.html (hunk 3 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/macos/Resources/snowflake-cli_welcome.html]: Updated all mac templates

File: scripts/packaging/macos/Resources/snowflake-cli_welcome.html (hunk 2 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/macos/Resources/snowflake-cli_welcome.html]: Updated content for SnowflakeCLI

File: scripts/packaging/build_darwin_package.sh (hunk 6 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/build_darwin_package.sh:41]: I used snowsql files as template.
I'm did not changed those entries on purpose.

File: scripts/packaging/win/wxs_builder.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-mraba) [reply scripts/packaging/win/wxs_builder.py]: The file was removed. I found another way to handle definition.

File: scripts/packaging/win/snowflake_cli.wxs (hunk 4 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/win/snowflake_cli.wxs]: Good point 👍

**Review by sfc-gh-astus — DISMISSED** [review]

File: scripts/packaging/build_darwin_package.sh (hunk 6 above)
Comment (sfc-gh-mraba) [reply scripts/packaging/build_darwin_package.sh]: 24h hours later we have a working build with `com.snowflake`

**Review by sfc-gh-turbaszek — APPROVED** [review]

**Review by sfc-gh-astus — APPROVED** [review]

**Pull request merged**